File: packages/expo/ios/ClerkExpoCoreConnection.swift
import ClerkKit
import Foundation
import UIKit

@MainActor final class ExpoCoreTransport: CoreTransport {
  var receive: (@MainActor (JSONValue) -> Void)?
  private var emit: ((String) -> Void)?
  init(emit: @escaping (String) -> Void) { self.emit = emit }
  func send(_ message: JSONValue) throws {
    guard let emit else { throw CoreError(code: "native_host_unavailable") }
    let data = try JSONEncoder().encode(message)
    guard data.count <= 16 * 1024 * 1024 else { throw CoreError(code: "message_too_large") }
    emit(String(decoding: data, as: UTF8.self))
  }
  func close() {
    emit = nil
    receive = nil
  }
}

private actor NoExpoClientStorage: CredentialStorage {
  func read() throws -> String? { throw CoreError(code: "client_owned_by_expo") }
  func write(_: String) throws { throw CoreError(code: "client_owned_by_expo") }
  func remove() throws { throw CoreError(code: "client_owned_by_expo") }
}

/// A native resource projection. This path never constructs JavaScriptCore or calls Clerk.connect.
@MainActor final class ClerkExpoCoreConnection {
  let id = UUID().uuidString
  let configuration: ClerkConfiguration
  let capabilities: AppleCapabilities
  let transport: ExpoCoreTransport
  let runtime: CoreRuntime
  private let authentication: AppleAuthentication
  private var requests: [String: Task<JSONValue, any Error>] = [:]
  private var closed = false

  init(publishableKey: String, emit: @escaping (String, String) -> Void) throws {
    guard
      let callback = Bundle.main.object(forInfoDictionaryKey: "ClerkNativeCallbackURL") as? String,
      let url = URL(string: callback)
    else { throw CoreError(code: "missing_native_callback_configuration") }
    configuration = try ClerkConfiguration(publishableKey: publishableKey, callbackURL: url)
    authentication = AppleAuthentication {
      UIApplication.shared.connectedScenes.compactMap { $0 as? UIWindowScene }
        .flatMap(\.windows).first(where: \.isKeyWindow) ?? UIWindow()
    }
    let config = configuration
    let authStorage = KeychainCredentialStorage(
      publishableKey: publishableKey, frontendAPI: config.frontendAPI, purpose: .magicLink)
    let biometrics = AppleBiometricCapabilities(
      publishableKey: publishableKey,
      credentials: KeychainCredentialStorage(
        publishableKey: publishableKey, frontendAPI: config.frontendAPI,
        purpose: .biometricCredentials),
      cleanup: KeychainCredentialStorage(
        publishableKey: publishableKey, frontendAPI: config.frontendAPI, purpose: .biometricCleanup)
    )
    capabilities = try AppleCapabilities(
      publishableKey: publishableKey, frontendAPI: config.frontendAPI,
      storage: NoExpoClientStorage(),
      browser: authentication.openBrowser, passkeys: authentication.credential,
      appleIdentity: authentication.appleIdentity, authStorage: authStorage, biometrics: biometrics
    )
    let connectionId = id
    transport = ExpoCoreTransport { message in emit(connectionId, message) }
    runtime = CoreRuntime(transport: transport)
  }

  var descriptor: [String: Any] {
    [
      "connectionId": id, "platform": "ios",
      "callbackUrl": configuration.callbackURL.absoluteString,
      "capabilities": capabilities.supported.filter {
        !["http", "storage", "timer", "random"].contains($0)
      },
    ]
  }

  func start() async throws -> Clerk {
    guard !closed else { throw CoreError(code: "native_host_unavailable") }
    let timeout = Task { @MainActor [weak self] in
      try await Task.sleep(for: .seconds(15))
      self?.close()
    }
    defer { timeout.cancel() }
    try await runtime.initialize(
      publishableKey: configuration.publishableKey, callbackURL: configuration.callbackURL,
      platform: "ios", capabilities: [])
    guard !closed, let clerk = try runtime.root("clerk", as: Clerk.self) else {
      throw CoreError(code: "native_host_unavailable")
    }
    return clerk
  }

  func receive(_ message: String) throws {
    guard !closed, message.utf8.count <= 16 * 1024 * 1024 else {
      throw CoreError(code: "native_host_unavailable")
    }
    transport.receive?(try JSONDecoder().decode(JSONValue.self, from: Data(message.utf8)))
  }

  func perform(id: String, capability: String, arguments: String) async throws -> String {
    guard !closed, requests[id] == nil, arguments.utf8.count <= 16 * 1024 * 1024,
      ["browser", "appleIdentity", "crypto.sha256"].contains(capability)
        || capability.hasPrefix("passkeys.") || capability.hasPrefix("authStorage.")
        || capability.hasPrefix("biometrics.")
    else { throw CoreError(code: "capability_unavailable") }
    let value = try JSONDecoder().decode(JSONValue.self, from: Data(arguments.utf8))
    let task = Task { @MainActor in try await capabilities.perform(capability, arguments: value) }
    requests[id] = task
    defer { requests.removeValue(forKey: id) }
    let result = try await task.value
    try Task.checkCancellation()
    guard !closed else { throw CoreError(code: "native_host_unavailable") }
    return String(decoding: try JSONEncoder().encode(result), as: UTF8.self)
  }

  func cancel(_ ids: [String]) {
    for id in ids { requests.removeValue(forKey: id)?.cancel() }
  }

  func close() {
    guard !closed else { return }
    closed = true
    for request in requests.values { request.cancel() }
    requests.removeAll()
    runtime.close()
  }
}

File: packages/expo/android/src/main/java/expo/modules/clerk/ClerkExpoCoreConnection.kt
package expo.modules.clerk

import android.app.Activity
import android.content.Context
import com.clerk.api.*
import java.util.UUID
import kotlinx.coroutines.*
import kotlinx.serialization.json.*

internal class ExpoCoreTransport(private var emit: ((String) -> Unit)?) : CoreTransport {
  override var receive: ((JsonElement) -> Unit)? = null
  override fun send(message: JsonElement) {
    val encoded = message.toString()
    if (encoded.length > 16 * 1024 * 1024) throw CoreException("message_too_large")
    (emit ?: throw CoreException("native_host_unavailable"))(encoded)
  }
  override fun close() { emit = null; receive = null }
}

private object NoExpoClientStorage : CredentialStorage {
  override suspend fun read(): String? = throw CoreException("client_owned_by_expo")
  override suspend fun write(value: String): Unit = throw CoreException("client_owned_by_expo")
  override suspend fun remove(): Unit = throw CoreException("client_owned_by_expo")
}

/** Projects Expo's existing owner. No QuickJS transport or Clerk.connect is constructed. */
internal class ClerkExpoCoreConnection(
  context: Context,
  publishableKey: String,
  activity: () -> Activity?,
  emit: (String, String) -> Unit,
) {
  val id = UUID.randomUUID().toString()
  val configuration = ClerkConfiguration(publishableKey, "${context.packageName}.clerk://oauth/callback")
  private val capabilities = AndroidCapabilities(
    publishableKey, configuration.frontendAPI, NoExpoClientStorage,
    activity, BrowserAuthentication(activity),
    AndroidCredentialStorage(context, publishableKey, purpose = AndroidCredentialStorage.Purpose.MAGIC_LINK),
    biometrics = AndroidBiometricCapabilities(
      context, publishableKey,
      AndroidCredentialStorage(context, publishableKey, purpose = AndroidCredentialStorage.Purpose.BIOMETRIC_CREDENTIALS),
      AndroidCredentialStorage(context, publishableKey, purpose = AndroidCredentialStorage.Purpose.BIOMETRIC_CLEANUP),
      activity,
    ),
  )
  private val transport = ExpoCoreTransport { emit(id, it) }
  val runtime = CoreRuntime(transport)
  private val scope = CoroutineScope(SupervisorJob() + Dispatchers.Main.immediate)
  private val requests = mutableMapOf<String, Deferred<JsonElement>>()
  private var closed = false
  val descriptor: Map<String, Any> get() = mapOf(
    "connectionId" to id, "platform" to "android", "callbackUrl" to configuration.callbackUrl,
    "capabilities" to (capabilities.supported - setOf("http", "storage", "timer", "random")).toList(),
  )

  suspend fun start(): Clerk {
    if (closed) throw CoreException("native_host_unavailable")
    try {
      withTimeout(15_000) { runtime.initialize(configuration.publishableKey, configuration.callbackUrl, "android", emptySet()) }
      if (closed) throw CoreException("native_host_unavailable")
      return runtime.resource(runtime.roots["clerk"] ?: throw CoreException("missing_clerk_root")) as Clerk
    } catch (error: Exception) { close(); throw error }
  }
  fun receive(message: String) {
    if (closed || message.length > 16 * 1024 * 1024) throw CoreException("native_host_unavailable")
    transport.receive?.invoke(Json.parseToJsonElement(message))
  }
  suspend fun perform(id: String, capability: String, arguments: String): String {
    if (closed || id in requests || arguments.length > 16 * 1024 * 1024 ||
      !(capability in setOf("browser", "googleIdentity", "crypto.sha256") || capability.startsWith("passkeys.") || capability.startsWith("authStorage.") || capability.startsWith("biometrics.")))
      throw CoreException("capability_unavailable")
    val args = Json.parseToJsonElement(arguments)
    val task = scope.async { capabilities.perform(capability, args) }
    requests[id] = task
    try { return task.await().toString() } finally { requests.remove(id) }
  }
  fun cancel(ids: List<String>) { ids.forEach { requests.remove(it)?.cancel() } }
  fun close() {
    if (closed) return
    closed = true
    requests.values.forEach { it.cancel() }
    requests.clear()
    scope.cancel()
    runtime.close()
  }
}
